### DISCUSSION TITLE
Socket error on client A4CF12FEEAA3, disconnecting. on tasmota 13.2.0

### DISCUSSION THREAD

**Discussion by RaphaelSquair** [arendst/Tasmota#20098] [Support request] [answered]
### PROBLEM DESCRIPTION
Having problem with stabilty of Tasmota device on a MQTT service

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): wemos d1 mini pro
- [x] Tasmota binary firmware version number used: 13.2.0
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: platformIO
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
{"Status":{"Module":0,"DeviceName":"DEFAULT","FriendlyName":["DEFAULT"],"Topic":"THERM/A4CF12FEEAA3","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":5,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
{"StatusPRM":{"Baudrate":127800,"SerialConfig":"7E2","GroupTopic":"THERM","OtaUrl":" ","RestartReason":"Power On","Uptime":"0T00:00:45","StartupUTC":"2023-11-24T16:05:46","Sleep":50,"CfgHolder":4617,"BootCount":19,"BCResetTime":"2023-11-10T12:26:29","SaveCount":59,"SaveAddress":"FFB000"}}
{"StatusFWR":{"Version":"13.2.0(tasmota)","BuildDateTime":"2023-10-20T12:54:54","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"458/699"}}
{"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["DARK","IOT_Squair"],"TelePeriod":300,"Resolution":"558180C0","SetOption":["0000200B","2805C80000000600003C5A00192800000000","000080C8","00006000","08004000","00000002"]}}
{"StatusMEM":{"ProgramSize":632,"Free":1412,"Heap":12,"ProgramFlashSize":16384,"FlashSize":16384,"FlashChipId":"18405E","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8780058B","04080001","0000004F","010013C0","00000008","00004080","00200000","04000020","00000000"],"Drivers":"1,2,3,4,8,9,10,12,38,50,62,68","Sensors":"1,2,3,5,53"}}
{"StatusNET":{"Hostname":"THERM-A4CF12FEEAA3-2723","IPAddress":"192.168.0.77","Gateway":"192.168.0.1","Subnetmask":"255.255.255.0","DNSServer1":"181.213.132.2","DNSServer2":"181.213.132.3","Mac":"A4:CF:12:FE:EA:A3","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":19.0}}
{"StatusMQT":{"MqttHost":"perfect-photographer.cloudmqtt.com","MqttPort":1883,"MqttClientMask":"Babel_%12X","MqttClient":"Babel_A4CF12FEEAA3","MqttUser":"cwcvnnpa","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":15,"SOCKET_TIMEOUT":4}}
{"StatusTIM":{"UTC":"2023-11-24T16:06:31","Local":"2023-11-24T13:06:31","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"-03:00","Sunrise":"04:12","Sunset":"13:00"}}
{"StatusSTS":{"Time":"2023-11-24T13:06:31","Uptime":"0T00:00:45","UptimeSec":45,"Heap":11,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":1,"Wifi":{"AP":1,"SSId":"DARK","BSSId":"E8:D2:FF:AA:5D:04","Channel":1,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:13"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
The error is intermittent, but it always start when the device is connected with AP 2 and is changed to AP1's wifi,  after the connection the device connects to the broker for 3 to 4 minutes and disconnects with error 
2023-11-24 16:09:24: Socket error on client A4CF12FEEAA3, disconnecting.

### EXPECTED BEHAVIOUR
It should maintain connection with the MQTT broker

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Probably it is hard to help here. We can not reproduce, and the problem happens sometimes for you too.
Using a cloud mqtt broker introduces extra complexity.

**Reply by barbudor** [reply]
@RaphaelSquair 
What is the frequency of problem ? once every 10 min ? once every hour ? once every day ?
What is your ping time to your broker ? Is it stable over long time ? => Over 10 minutes I see quite stable 84ms from France over a fiber connection
Which contract do you have with cloudmqtt ?
The cheapest "Humble Hedgehog" clearly state "variable performance"

Some commands may be helpful when dealing with a cloud broker:
- MqttWifiTimeout : increasing to 1000 or more can solve creating the connection 
- MqttTimeout : could be increased to in order to avoid disconnection in case of internet congestion
- MqttKeepAlive may help but I'm not sure if it should be increased or decreased

Additional comments:
- `"MqttPort":1883` : are you really using unencrypted connection over the internet ?
- `"Topic":"THERM/A4CF12FEEAA3"` : You should not have a `/` in your device topic. Topic is supposed to represent a single level. Either use a `-` or `_`, or use FullTopic
As per doc: 
![image](https://github.com/arendst/Tasmota/assets/2996042/29914db8-7aa8-4f21-b88d-496c9707fac5)

**Reply by sfromis** [reply]
Maybe the cloud broker does not "like" clients disconnecting and reconnecting. It might be worth a shot to keep the device on the same AP instead of this switching.

**Reply by barbudor** [reply]
Is the IP the same when swiching from one Ap to the other AP ?
If the IP is changing then it is normal that the connection gets broken

Getting Tasmota log in weblog 3 at that moment could be usefull

**Reply by sfromis** [reply]
Even assuming same IP address, Tasmota would still be reconnecting to the broker after reconnecting to Wifi.

**Reply by barbudor** [reply]
Yes but my understanding is that it reconnects for a couple of minutes and then disconnects

That's what I would like to see in the logs

**Reply by sfromis** [reply]
And also broker log including time of Tasmota switching AP, and when it disconnected again.

**Reply by RaphaelSquair** [reply]
Hi Guys, thanks for the attention, if turn weblog 3 on and try to capture this behavior. I need to Change from AP2 to AP1 because i use the AP2 with a maintenance Ssid and the AP1 with the normal use ssid. thats for safety, if the first ssid dont connect , i can create a hotspot on my smartphone to see what is happening.

**Comment by RaphaelSquair** [comment]
Hey guys anyone could shed a light in this problem? i1ve had others boards with the same problem, there is a command to clear the ip? or any way to send a flush stack so it connect to the ssid1 with a clean stack? or only a restart will do?

**Comment by RaphaelSquair** [comment, marked as answer]
Just to answer any one that may have this problems or any other related with connectivity: 
Measure both the latency or ping to the dns server and the mqtt server you are using and apply at least 5 times the amount of latency  using the following commands.

MqttTimeout  10 ( i've used 10 s because in  the worst case scenario i've seen using a external mqtt server was almost 2 sec using a 3g modem in a isolated site)

MqttWifiTimeout 2000 (I've used 2000 because i'm using my device with a external router and in a external site with lots of interference and high traffic of cars and people where packet loss is relative common.)

I hope these tips help someone in the future

my reference is this comment (https://github.com/arendst/Tasmota/issues/12501#issuecomment-881419663) that @arendst posted some time ago, it was very helpfull in debuggin my problem, it look like most of the wifi routers and 3g modems in brazil apears to introduce significant latency in the loop.
